cpp/src/parquet/arrow/arrow_reader_writer_test.cc

```diff
@@ -5146,5 +5158,361 @@ TEST(TestArrowReadWrite, FuzzReader) {
+    ASSERT_FALSE(c2_column_index->null_pages()[0]);
+    ASSERT_EQ(BoundaryOrder::Ascending, c2_column_index->boundary_order());
+    ASSERT_EQ(c2_min_values[rg], c2_column_index->encoded_min_values()[0]);
+    ASSERT_EQ(c2_max_values[rg], c2_column_index->encoded_max_values()[0]);
+    ASSERT_TRUE(c2_column_index->has_null_counts());
+    ASSERT_EQ(c2_null_counts[rg], c2_column_index->null_counts()[0]);
+
+    // Verify offset index.
+    for (int c = 0; c < metadata->num_columns(); ++c) {
+      auto offset_index = row_group_index_reader->GetOffsetIndex(c);
+      ASSERT_NE(offset_index, nullptr);
+      ASSERT_EQ(num_pages, offset_index->page_locations().size());
+      ASSERT_EQ(0, offset_index->page_locations()[0].first_row_index);
+      ASSERT_NE(0, offset_index->page_locations()[0].offset);
+      ASSERT_NE(0, offset_index->page_locations()[0].compressed_page_size);
+    }
+  }
+}
+
+TEST(TestArrowReadWrite, WriteReadPageIndexRoundTripWithLargeStatsDropped) {
+  // Enable page index to the writer.
+  auto writer_properties = WriterProperties::Builder()
+                               .enable_write_page_index()
+                               ->max_row_group_length(1) /* write single-row row group */
+                               ->max_statistics_size(20) /* drop stats larger than it */
+                               ->build();
+  auto arrow_writer_properties = default_arrow_writer_properties();
+  auto pool = ::arrow::default_memory_pool();
+  auto sink = CreateOutputStream();
+  auto schema = ::arrow::schema({::arrow::field("c0", ::arrow::utf8())});
+  std::shared_ptr<SchemaDescriptor> parquet_schema;
+  ASSERT_OK_NO_THROW(ToParquetSchema(schema.get(), *writer_properties,
+                                     *arrow_writer_properties, &parquet_schema));
+  auto schema_node = std::static_pointer_cast<GroupNode>(parquet_schema->schema_root());
+
+  // Create writer to write data via Table.
+  auto writer = ParquetFileWriter::Open(sink, schema_node, writer_properties);
+  std::unique_ptr<FileWriter> arrow_writer;
+  ASSERT_OK(FileWriter::Make(pool, std::move(writer), schema, arrow_writer_properties,
+                             &arrow_writer));
+  auto table = ::arrow::TableFromJSON(schema, {R"([
+      ["short_string"],
+      ["very_large_string_to_drop_stats"]
+    ])"});
+  ASSERT_OK_NO_THROW(arrow_writer->WriteTable(*table));
+  ASSERT_OK_NO_THROW(arrow_writer->Close());
+  ASSERT_OK_AND_ASSIGN(auto buffer, sink->Finish());
+
+  // Create reader to read page index.
+  auto read_properties = default_arrow_reader_properties();
+  auto reader = ParquetFileReader::Open(std::make_shared<BufferReader>(buffer));
+  auto metadata = reader->metadata();
+  ASSERT_EQ(2, metadata->num_row_groups());
+
+  // Make sure page index reader is not null.
+  auto page_index_reader = reader->GetPageIndexReader();
+  ASSERT_NE(page_index_reader, nullptr);
+
+  const size_t num_pages = 1;
+  for (int rg = 0; rg < metadata->num_row_groups(); ++rg) {
```

Comment:
Are all the loops necessary in this test?  it seems like there should only be two row-groups and one column?

Given the repetiveness of these tests, is there someway to make the logic more readable (some sort of custom gmock matcher?)

Reply:
or maybe helper functions to consolidate elements into a vector and use EXPECT_THAT assertions over array elements?